florestad/src/json_rpc/blockchain.rs

```diff
@@ -44,6 +46,138 @@ impl<Blockchain: RpcChain> RpcImpl<Blockchain> {
+    /// `timestamp` has an alias, 0 will directly refer to 1231006505, the time of the genesis block  
+    pub async fn get_block_height_by_timestamp(&self, timestamp: u32) -> Result<u32, JsonRpcError> {
+        if timestamp == 0 {
+            return Ok(timestamp);
+        };
+
+        const TEN_MINUTES: u32 = 600;
+        const GENESIS_TIMESTAMP: u32 = 1231006505; // The u32 representation of 03 Jan 2009, 18:15:05. The salvation of human kind.
+
+        if timestamp <= GENESIS_TIMESTAMP {
+            return Err(JsonRpcError::InvalidTimestamp);
+        }
+
+        let mut height_target = (timestamp - GENESIS_TIMESTAMP) / TEN_MINUTES;
+
+        // Find the first block at or below the target timestamp
+        while height_target > 0 {
+            let block_hash = match self.chain.get_block_hash(height_target) {
+                Ok(hash) => hash,
+                Err(_) => {
+                    height_target -= 1;
+                    continue;
+                }
+            };
+
+            let block_header = match self.chain.get_block_header(&block_hash) {
+                Ok(header) => header,
+                Err(_) => {
+                    height_target -= 1;
+                    continue;
+                }
+            };
+
+            let diff: i32 = (timestamp - block_header.time) as i32;
+            let abs_diff = diff.unsigned_abs();
+            if diff.is_negative() {
+                // we got a block thats too high
+                if abs_diff > TEN_MINUTES {
+                    // When theres more than TEN_MINUTES of difference we can use the diff to have
+                    // a more precise guess of the block
+                    height_target = height_target.saturating_sub(TEN_MINUTES.div_ceil(abs_diff));
+                    continue;
+                }
+            }
+
+            if diff.is_positive() {
+                // we got a block thats too high
+                if abs_diff > TEN_MINUTES {
+                    // When theres more than TEN_MINUTES of difference we can use the diff to have
+                    // a more precise guess of the block
+                    height_target.add_assign(TEN_MINUTES.div_ceil(abs_diff));
+                    continue;
+                }
+            }
+            // at this point the expected diff is less than 10 minutes from the timestamp so we break expecting to find the closes
+            // block in a 2 hours spam
+            break;
+        }
+
+        // This block spam is the amount of blocks we search for around the height we
+        // got to find the closest one.
+        //
+        // the rationale behind 24 is that a block can have a valid timestamp in a 2 hour spam
+        // it can be 2 hours in the future and 2 hours in the past and this 4 hour spam should
+        // contain around 24 blocks. This is more a stop point in case the given timestamp is
+        // not precise and a block doesnt contain the exact timestamp were trying to find.
```

Comment:
You can have way more than 24 blocks in two hours, just look at testnet{3, 4} for an example. There's not limit to how many blocks we may need to look for.

The ideal strategy here is a bisection. The code above doesn't work for testenets, as blocks are found **way** quicker than 10 minutes